bundles/org.openhab.binding.teslapowerwall/src/main/java/org/openhab/binding/teslapowerwall/internal/TeslaPowerwallHandler.java

```diff
@@ -0,0 +1,179 @@
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(TeslaPowerwallConfiguration.class);
+        logger.debug("config.hostname = {}, refresh = {}", config.hostname, config.refresh);
+        if (config.hostname.isBlank()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Hostname/IP address must be set");
+        } else {
+            webTargets = new TeslaPowerwallWebTargets(config.hostname);
+            refreshInterval = config.refresh;
+            updateStatus(ThingStatus.UNKNOWN);
+            schedulePoll();
+        }
+    }
+
+    @Override
+    public void dispose() {
+        super.dispose();
+        stopPoll();
+    }
+
+    private void schedulePoll() {
+        ScheduledFuture<?> pollFuture = this.pollFuture;
+        if (pollFuture != null) {
+            pollFuture.cancel(false);
+        }
+        logger.debug("Scheduling poll for 1 second out, then every {} s", refreshInterval);
+        this.pollFuture = scheduler.scheduleWithFixedDelay(this::poll, 1, refreshInterval, TimeUnit.SECONDS);
+    }
+
+    private void poll() {
+        try {
+            logger.debug("Polling for state");
+            pollStatus();
+        } catch (IOException e) {
+            logger.debug("Could not connect to Tesla Powerwall", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        } catch (RuntimeException e) {
+            logger.warn("Unexpected error connecting to Tesla Powerwall", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        }
+    }
+
+    private void stopPoll() {
+        final Future<?> future = pollFuture;
+        if (future != null && !future.isCancelled()) {
+            future.cancel(true);
+            pollFuture = null;
+        }
+    }
+
+    private void pollStatus() throws IOException {
+        if (!(config.email.isBlank() && config.password.isBlank()))
+            try {
+                if (token.isEmpty()) {
+                    token = webTargets.getToken(config.email, config.password);
+                }
+                Operations operations = webTargets.getOperations(token);
+                updateState(TeslaPowerwallBindingConstants.CHANNEL_TESLAPOWERWALL_MODE,
+                        new StringType(operations.mode));
+                updateState(TeslaPowerwallBindingConstants.CHANNEL_TESLAPOWERWALL_RESERVE,
+                        new QuantityType<>(((operations.reserve / 0.95) - (5 / 0.95)), Units.PERCENT));
+
+                BatterySOE batterySOE = webTargets.getBatterySOE(token);
+                GridStatus gridStatus = webTargets.getGridStatus(token);
+                SystemStatus systemStatus = webTargets.getSystemStatus(token);
+                MeterAggregates meterAggregates = webTargets.getMeterAggregates(token);
+
+                updateStatus(ThingStatus.ONLINE);
```

Comment:
```suggestion
            try {
                if (token.isEmpty()) {
                    token = webTargets.getToken(config.email, config.password);
                }
                Operations operations = webTargets.getOperations(token);
                BatterySOE batterySOE = webTargets.getBatterySOE(token);
                GridStatus gridStatus = webTargets.getGridStatus(token);
                SystemStatus systemStatus = webTargets.getSystemStatus(token);
                MeterAggregates meterAggregates = webTargets.getMeterAggregates(token);
                updateStatus(ThingStatus.ONLINE);
                } catch (TeslaPowerwallCommunicationException e) {
                logger.debug("Unexpected error connecting to Tesla Powerwall", e);
                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
                return;
                }
                
                updateState(TeslaPowerwallBindingConstants.CHANNEL_TESLAPOWERWALL_MODE,
                        new StringType(operations.mode));
                updateState(TeslaPowerwallBindingConstants.CHANNEL_TESLAPOWERWALL_RESERVE,
                        new QuantityType<>(((operations.reserve / 0.95) - (5 / 0.95)), Units.PERCENT));
```

This suggestion restructures/seperates the calls, catch block and update logic, might need spotless

Reply:
I tried this but got errors like: 
`[ERROR] /home/psmedley/openhab-addons/bundles/org.openhab.binding.teslapowerwall/src/main/java/org/openhab/binding/teslapowerwall/internal/TeslaPowerwallHandler.java:[128,96] operations cannot be resolved to a variable
`

Reply:
Updated the example, if you declare the cars before the block it should be fine.

Reply:
I got this working,  but had to do:
`Operations operations = null;` to avoid errors "The local variable operations may not have been initialized"

With this, I get warnings: "Potential null pointer access: The variable operations may be null at this location" so not entirely happy with this...

Reply:
OK I missed the 'return;' in the exception loop. Meanwhile,  I worked it out by myself and maybe learned something.